src/reconcile.py

```diff
@@ -0,0 +1,274 @@
+        )
+
+    # This is an edge case that can't actually occur because table_row.is_group is based on
+    # whether the navlink link is None so this case would have been caught in the group case
+    if table_row.navlink.link is None:  # pragma: no cover
+        raise exceptions.ReconcilliationError(
+            f"internal error, expecting link on table row, {table_row=!r}"
+        )
+    return types_.DeleteAction(
+        action=types_.Action.DELETE,
+        level=table_row.level,
+        path=table_row.path,
+        navlink=table_row.navlink,
+        content=discourse.retrieve_topic(url=table_row.navlink.link),
+    )
+
+
+def _calculate_action(
+    path_info: types_.PathInfo | None, table_row: types_.TableRow | None, discourse: Discourse
+) -> types_.AnyAction:
+    """Calculate the required action for a page.
+
+    Args:
+        path_info: Information about the local documentation file.
+        table_row: A row from the navigation table.
+        discourse: A client to the documentation server.
+
+    Raises:
+        ReconcilliationError: if both path_info and table_row are None.
+    """
+    if path_info is None and table_row is None:
+        raise exceptions.ReconcilliationError(
+            "internal error, both path info and table row are None"
+        )
+    if path_info is not None and table_row is None:
+        return _local_only(path_info=path_info)
+    if path_info is None and table_row is not None:
+        return _server_only(table_row=table_row, discourse=discourse)
+    if path_info is not None and table_row is not None:
+        return _local_and_server(path_info=path_info, table_row=table_row, discourse=discourse)
+
+    # Something weird has happened since all cases should already be covered
+    raise exceptions.ReconcilliationError("internal error")  # pragma: no cover
+
+
+def run(
+    path_infos: typing.Iterable[types_.PathInfo],
+    table_rows: typing.Iterable[types_.TableRow],
+    discourse: Discourse,
+) -> typing.Iterator[types_.AnyAction]:
+    """Reconcile differences between the docs directory and documentation server.
+
+    Preserves the order of path_infos although does not for items only in table_rows.
+
+    Args:
+        path_infos: Information about the local documentation files.
+        table_rows: Rows from the navigation table.
+        discourse: A client to the documentation server.
+
+    Returns:
+        The actions required to reconcile differences between the documentation server and local
+        files.
+    """
+    path_info_lookup: types_.PathInfoLookup = {
+        (path_info.level, path_info.table_path): path_info for path_info in path_infos
+    }
+    table_row_lookup: types_.TableRowLookup = {
+        (table_row.level, table_row.path): table_row for table_row in table_rows
+    }
+
+    # Need to process items only on the server last
+    keys = itertools.chain(
+        path_info_lookup.keys(), table_row_lookup.keys() - path_info_lookup.keys()
+    )
```

Comment:
I'll explain what is happening and see if you have any feedback after that. I'll also include what I write here in the comment explaining the line.

This function needs to match files and directories locally to items on the navigation table on the server knowing that there may be cases that are not matched. The navigation table relies on the order that items are displayed to figure out the hierarchy/ page grouping (this is not a design choice of this action but how the documentation is interpreted by charmhub). The `path_infos` have been ordered to ensure that the hierarchy will be calculated correctly by the server when the new navigation table is generated. This is why an overall set is not possible since it won't preserve the order of items. (Note that `path_info_lookup` will preserve the insertion order as dictionaries are now guaranteed to preserve insertion order so the conversion to dictionaries does not destroy the ordering of the `path_infos`.) Additionally, this line is already as efficient as using an overall set whilst guaranteeing ordering is preserved because what is returned by `dict.keys()` already behaves like a set. If you do an inspection of what `table_row_lookup.keys() - path_info_lookup.keys()` returns, it is actually a set.

The next question might be that this means for `table_row_lookup.keys() - path_info_lookup.keys()` the order is not preserved, and that is true but doesn't matter since those items are the items that are only on the server, i.e., those keys will just result in delete actions which have no effect on the navigation table that is generated and hence ordering for them doesn't matter.

Does that make sense?

Reply:
This has now been pushed to this PR

Reply:
I understand the reasoning behind, but that's a looot of hidden logic behind few lines. The comment at least is important to explain all this.
Isn't there a way to recompute the proper ordering when the navigation table is generated ? It, at least, shows that there is something "wrong" in the design of the navigation table, that should include an ordering attribute for programmatic management. Perhaps a suggestion to do to the documentation team ?

Reply:
Marking as resolved for now, please let me know if there are any further issues on this topic